### PR TITLE
Update summary for BinaryData in mgmt plane

### LINKED ISSUE #2447: [P0][Bug] The serialization will fail if we change a string type to BinaryData
We changed the type of string to BinaryData at some times. Here is one example of [Cdn](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/cdn/Azure.ResourceManager.Cdn/src/Generated/Models/WafPolicySettings.cs#L44). Unfortunately, currently our serialization code will fail in such case because we will try to [parse](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/cdn/Azure.ResourceManager.Cdn/src/Generated/Models/WafPolicySettings.Serialization.cs#L51-L55) the string, but it isn't a valid json value.

Considering that this has impact on the packages with stable version, I think we should fix it ASAP.

### PR DESCRIPTION
**Pull request by Yao725**
# Description
Fixes #2447 

The original implementation of BinaryData was intended for `AnyObject` which can have any structure but must be a valid json.  It was expanded to include `Any` but when that happened we didn't adjust the serialization to handle the additional cases.

This PR is exploring options on how to make `Any` work in the serialization code.  The deserialization should remain untouched as no interpretation of the data happens during deserialization we simply store the raw bytes into the BinaryData and let the customer choose how to interpret the data.

For serialization we need to inspect the BinaryData and determine if it should be written as raw bytes (`json`) or as one of the primitive types `null`, `bool`, `number`, `string`.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 1]
```diff
@@ -140,5 +140,26 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
                     throw new NotSupportedException("Not supported type " + value.GetType());
             }
         }
+
+        public static void WriteBinaryDataValue(this Utf8JsonWriter writer, BinaryData value)
+        {
+            string inputValue = value.ToString();
+            // First we check if it is possible to serialize the value as json value.
+            if (inputValue.StartsWith("{", StringComparison.InvariantCultureIgnoreCase) && (inputValue.EndsWith("}", StringComparison.InvariantCultureIgnoreCase) || inputValue.EndsWith($"}}{Environment.NewLine}", StringComparison.InvariantCultureIgnoreCase)) ||
+                inputValue.StartsWith("[", StringComparison.InvariantCultureIgnoreCase) && (inputValue.EndsWith("]", StringComparison.InvariantCultureIgnoreCase) || inputValue.EndsWith($"]{Environment.NewLine}", StringComparison.InvariantCultureIgnoreCase)) ||
```
Comment (ArthurMa1978) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:149]: consider to have a perf test

Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:149]: I tested the performance using implementation with try-catch and without try-catch respectively and below is the result.

- When the BinaryData is a json object

|         Method |           Job |       Runtime |       Mean |    Error |   StdDev |     Median |        Min |        Max |  Gen 0 | Allocated |
|--------------- |-------------- |-------------- |-----------:|---------:|---------:|-----------:|-----------:|-----------:|-------:|----------:|
| TestWithoutTry |      .NET 6.0 |      .NET 6.0 |   920.7 ns |  2.85 ns |  2.53 ns |   920.4 ns |   917.1 ns |   924.7 ns | 0.1574 |      1 KB |
|    TestWithTry |      .NET 6.0 |      .NET 6.0 |   952.4 ns | 10.76 ns | 10.07 ns |   954.7 ns |   937.2 ns |   968.7 ns | 0.1574 |      1 KB |
| TestWithoutTry | .NET Core 3.1 | .NET Core 3.1 | 1,576.0 ns | 23.19 ns | 21.69 ns | 1,566.8 ns | 1,551.4 ns | 1,621.7 ns | 0.1583 |      1 KB |
|    TestWithTry | .NET Core 3.1 | .NET Core 3.1 | 1,618.4 ns | 33.16 ns | 94.60 ns | 1,568.3 ns | 1,541.2 ns | 1,862.0 ns | 0.1583 |      1 KB |

- When the BinaryData is a string and we correctly treat it as string

|         Method |           Job |       Runtime |     Mean |   Error |  StdDev |      Min |      Max |  Gen 0 |  Gen 1 | Allocated |
|--------------- |-------------- |-------------- |---------:|--------:|--------:|---------:|---------:|-------:|-------:|----------:|
| TestWithoutTry |      .NET 6.0 |      .NET 6.0 | 201.0 ns | 0.62 ns | 0.55 ns | 200.1 ns | 202.1 ns | 0.1013 | 0.0002 |     848 B |
|    TestWithTry |      .NET 6.0 |      .NET 6.0 | 201.5 ns | 0.75 ns | 0.70 ns | 200.6 ns | 203.2 ns | 0.1013 | 0.0002 |     848 B |
| TestWithoutTry | .NET Core 3.1 | .NET Core 3.1 | 312.0 ns | 1.46 ns | 1.30 ns | 309.8 ns | 314.2 ns | 0.1020 |      - |     856 B |
|    TestWithTry | .NET Core 3.1 | .NET Core 3.1 | 333.8 ns | 3.62 ns | 3.21 ns | 328.7 ns | 338.1 ns | 0.1020 |      - |     856 B |

Based on above results, we can find that if we add a try-catch in this method, there will a little performance problem(however, sometimes the time are even very close to each other), but it can ensure that we could correctly handle the situation that a binary data is actually a string but we see it as a json. To be honest, I think this kind of case will be very rare, but we don't have the confidence to say it will never exist, therefore I think it's worth sacrificing a little performance here.

**Review by annelo-msft — COMMENTED** [review]
Could you share what issue this is addressing?  I'm curious what motivates this addition.

**Comment by Yao725** [comment]
> Could you share what issue this is addressing? I'm curious what motivates this addition.

Oh, my bad, I forgot to link the [issue](https://github.com/Azure/autorest.csharp/issues/2447) in the PR description. The context is that previously we changed the type from object to BinaryData and will parse it as we regard it as a valid json value. And in the process of releasing some management plane SDKs, we also have changed the type of some properties from string to BinaryData. At this time, the original serialization code will not work. Therefore, we want to have this change.

**Comment by annelo-msft** [comment]
Great, thank you for the explanation!

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 2]
```diff
@@ -140,5 +140,33 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
                     throw new NotSupportedException("Not supported type " + value.GetType());
             }
         }
+
+        public static void WriteBinaryDataValue(this Utf8JsonWriter writer, BinaryData value)
```
Comment (AlexanderSher) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:144]: Why not make it a case in `WriteObjectValue`?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 1 above)
Comment (AlexanderSher) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:149]: How big was the underlying byte array?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 3]
```diff
@@ -140,5 +140,33 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+            {
+                try
+                {
+#if NET6_0_OR_GREATER
+				    writer.WriteRawValue(value);
+#else
+                    JsonSerializer.Serialize(writer, JsonDocument.Parse(inputValue).RootElement);
+#endif
+                }
+                catch
```
Comment (AlexanderSher) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:161]: Please specify the exception you're catching here.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 4]
```diff
@@ -140,5 +140,33 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
                     throw new NotSupportedException("Not supported type " + value.GetType());
             }
         }
+
+        public static void WriteBinaryDataValue(this Utf8JsonWriter writer, BinaryData value)
+        {
+            string inputValue = value.ToString();
```
Comment (AlexanderSher) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:146]: I don't like that we need to create a copy of the `BinaryData` content. Can we cast `BinaryData` to `Span<byte>` and then use [UTF8Encoding.GetChars](https://docs.microsoft.com/en-us/dotnet/api/system.text.utf8encoding.getchars?view=net-6.0)?

Comment (AlexanderSher) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:146]: Or maybe even use `Utf8JsonReader` to validate the span?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 2 above)
Comment (ArthurMa1978) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:144]: Object is Json Object, but here the BinaryData can ben raw string or raw binary

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 1 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:151]: 1. Vaule: `{"foo":"bar","numbers":[1,2,3,4]}`, Length: 33
2. Vaule: `abcdef`, Length: 6

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 2 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:145]: I think there are two reasons that we don't want to make it a case in `WriteObjectValue`:
1. In mgmt plane, we explicitly know that the input parameter is a BinaryData, so we don't need the type check in `WriteObjectValue`
2. `WriteObjectValue` already has a case about `BinaryData`, which writes the value as Base64 encoded json string, I don't think we want to break current algorithm.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 3 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: specified.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 4 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: I change it to ReadOnlySpan<char> using `GetChars`.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 5]
```diff
@@ -140,5 +141,34 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
                     throw new NotSupportedException("Not supported type " + value.GetType());
             }
         }
+
+        public static void WriteBinaryDataValue(this Utf8JsonWriter writer, BinaryData value)
+        {
+            ReadOnlySpan<char> inputValue = UTF8Encoding.UTF8.GetChars(value.ToArray()).AsSpan();
+            // Check if it is possible to serialize the value as json value.
+            // Possible match kinds: Json Object, Json Array, Json value false, Json value true, Json number.
+            if (inputValue.StartsWith("{".AsSpan(), StringComparison.InvariantCultureIgnoreCase) && (inputValue.EndsWith("}".AsSpan(), StringComparison.InvariantCultureIgnoreCase) || inputValue.EndsWith($"}}{Environment.NewLine}".AsSpan(), StringComparison.InvariantCultureIgnoreCase)) ||
```
Comment (m-nash) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:150]: lets use index comparison vs startswith

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 6]
```diff
@@ -140,5 +141,34 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+                inputValue.StartsWith("[".AsSpan(), StringComparison.InvariantCultureIgnoreCase) && (inputValue.EndsWith("]".AsSpan(), StringComparison.InvariantCultureIgnoreCase) || inputValue.EndsWith($"]{Environment.NewLine}".AsSpan(), StringComparison.InvariantCultureIgnoreCase)) ||
```
Comment (m-nash) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:151]: instead of endswith `}Newline` lets start at the end and move inwards until we no longer have whitespace then use that as the last character

File: test/AutoRest.Shared.Tests/WriterExtensionTests.cs [hunk 7]
```diff
@@ -37,6 +38,20 @@ public class WriterExtensionTests
             },
         };
 
+        private static object[] BinaryDataCases =
+        {
+            new object[] { BinaryData.FromString("abcdefg"), "\"abcdefg\"" },
+            new object[] { BinaryData.FromString("123"), "123" },
```
Comment (m-nash) [inline test/AutoRest.Shared.Tests/WriterExtensionTests.cs:44]: lets add test cases around .FromObject(123)

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 8]
```diff
@@ -140,5 +141,34 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+#endif
+                }
+                catch (JsonException)
+                {
+                    writer.WriteStringValue(inputValue);
+                }
+            }
+            else
+            {
+                writer.WriteStringValue(inputValue);
```
Comment (m-nash) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:170]: If we are always writing string value do we ever want to be able to serialize `"foo": 51`?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 5 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: updated

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 6 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: I have added a trim logic and we don't need this NewLine check anymore.

File: test/AutoRest.Shared.Tests/WriterExtensionTests.cs (hunk 7 above)
Comment (Yao725) [reply test/AutoRest.Shared.Tests/WriterExtensionTests.cs:44]: Added

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 8 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:189]: Our logic here will check if the first byte and last byte is between `0` and `9`, if it is, we will parse it as json number. There fore if you use an integer to instantiate the BinaryData, no matter you do it by `BinaryData.FromString("51")` or `BinaryData.FromObjectAsJson(51)`, the result will be `"foo": 51`

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 9]
```diff
@@ -140,5 +141,53 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
                     throw new NotSupportedException("Not supported type " + value.GetType());
             }
         }
+
+        public static void WriteBinaryDataValue(this Utf8JsonWriter writer, BinaryData value)
+        {
+            byte[] inputValue = value.ToArray();
+
+            // We will remove the trailing occurrences of the specified characters including: TAB, LF, space.
+            byte[] trimBytes = { 9, 10, 32 };
```
Comment (m-nash) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:150]: can we use `char.IsWhiteSpace` here?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 10]
```diff
@@ -140,5 +141,53 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+                }
+                if (i == trimBytes.Length)
+                    break;
+            }
+
+            // Check if it is possible to serialize the value as json value.
+            // Possible match kinds: Json Object, Json Array, Json value false, Json value true, Json number.
+            // Our logic here is that, for example, we check if the first and the last bytes are '{' and '}', respectively.
+            if (end >= 0 && (
+                inputValue[0] == 123 && inputValue[end] == 125 ||
```
Comment (m-nash) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:168]: can we add constants so we know what the ints stand for?

File: test/AutoRest.Shared.Tests/WriterExtensionTests.cs [hunk 11]
```diff
@@ -6,6 +6,7 @@
 using System.IO;
 using System.Text.Json;
 using NUnit.Framework;
+using JsonObject = System.Collections.Generic.Dictionary<string, object>;
```
Comment (m-nash) [inline test/AutoRest.Shared.Tests/WriterExtensionTests.cs:9]: Can we add test cases that test the boundaries where we start with '{' but don't end with '}' as well as the other combos we are looking for
Add test cases with various trailing whitespace
Cases with all whitespace

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 12]
```diff
@@ -140,5 +141,53 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
                     throw new NotSupportedException("Not supported type " + value.GetType());
             }
         }
+
+        public static void WriteBinaryDataValue(this Utf8JsonWriter writer, BinaryData value)
+        {
+            byte[] inputValue = value.ToArray();
```
Comment (m-nash) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:147]: Does this create a copy?  If so is there a way to access the raw bytes without a copy?

File: test/AutoRest.Shared.Tests/WriterExtensionTests.cs [hunk 13]
```diff
@@ -87,6 +103,17 @@ public static void WriteObjectValueNullIUtf8JsonSerializable ()
...
 
+        [Test, TestCaseSource("BinaryDataCases")]
+        public static void WriteBinaryDataValue(BinaryData value, string expectedValue)
+        {
+            using MemoryStream stream = new MemoryStream();
+            using Utf8JsonWriter writer = new Utf8JsonWriter(stream);
+            writer.WriteBinaryDataValue(value);
+
+            writer.Flush();
+            Assert.AreEqual(expectedValue, System.Text.Encoding.UTF8.GetString(stream.ToArray()));
```
Comment (m-nash) [inline test/AutoRest.Shared.Tests/WriterExtensionTests.cs:114]: Can we add round trip tests to serialize, then deserialize and validate the original byte array is equal to the new byte array.

I don't know if it will make a difference but can we also validate these cases while they are a property of a larger object?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 14]
```diff
@@ -140,5 +141,53 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+                    break;
+            }
+
+            // Check if it is possible to serialize the value as json value.
+            // Possible match kinds: Json Object, Json Array, Json value false, Json value true, Json number.
+            // Our logic here is that, for example, we check if the first and the last bytes are '{' and '}', respectively.
+            if (end >= 0 && (
+                inputValue[0] == 123 && inputValue[end] == 125 ||
+                inputValue[0] == 91 && inputValue[end] == 93 ||
+                inputValue[0] == 116 && inputValue[end] == 101 ||
```
Comment (m-nash) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:170]: we are saying that if the raw bytes = 't', ..., 'e' we will treat this as json?  Lets add a test case that sets the bytes equal to `terse`.

I am still not fully sold on the true, false, and number checks here.  If we send them to write as json 456 comes out as "456" according to the test case [here](https://github.com/Azure/autorest.csharp/pull/2544/files#diff-ede2da49b9fb105e853158da8dfe79723aa96156ae93c3577a7b0485fdf24c39R45)?  If we send it to write as string doesn't 456 also become "456"?

If we want to actually do anything here for true / false lets check length then validate the full word.

**Review by m-nash — CHANGES_REQUESTED** [review]

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 15]
```diff
@@ -140,5 +141,53 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
                     throw new NotSupportedException("Not supported type " + value.GetType());
             }
         }
+
+        public static void WriteBinaryDataValue(this Utf8JsonWriter writer, BinaryData value)
```
Comment (annelo-msft) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:145]: It would be good to get the Azure.Core v-team's eyes on this change per the change to Azure.Core shared source.  @jsquire, @JoshLove-msft, @christothes

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 9 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: changed

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 10 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: changed

File: test/AutoRest.Shared.Tests/WriterExtensionTests.cs (hunk 11 above)
Comment (Yao725) [reply test/AutoRest.Shared.Tests/WriterExtensionTests.cs:9]: added

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 12 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: Change to use ReadOnlySpan<byte> to avoid copying

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 14 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: I have added the test case. 

If we write 456 as string, the result will be "\"456\"", not "456". 

I understand the we can check the length and validate the full word for true/false, but what I want to say is, we are using the first and last byte to check if the binary data is likely to be a valid json and we should keep this pattern for the different json kinds. In 90% of the cases, we will pass this check and the json parse will succeed, that's what we are expecting. And we are clear that there are still some cases that the binary data can pass this check but it can't be parsed, therefore we will catch the exception and write it as string.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 16]
```diff
@@ -140,5 +141,52 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+            int end = inputValue.Length - 1;
+            while (end >= start && char.IsWhiteSpace(Convert.ToChar(inputValue[end])))
+            {
+                end--;
+            }
+            int len = end - start + 1;
+
+            // Check if it is possible to serialize the value as json value.
+            // Possible match kinds: Json Object, Json Array, Json string, Json value true, Json value false, Json number.
+            // Our logic here is that, for example, we check if the first and the last bytes are '{' and '}', respectively.
```
Comment (christothes) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:164]: Is this heuristic enough? What about values like `{   this{ is { invalid { josn }` ?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 17]
```diff
@@ -140,5 +141,52 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+            if (len > 0 && (
```
Comment (jsquire) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:165]: The allow-list approach makes me uncomfortable.   For example, this would consider `null` invalid and thus rule out valid patterns like:

```json
"someKey" : null
```

What problem are we trying to solve here?   Why do we need this level of validation rather than just writing out the bytes?   `Utf8JsonWriter` will validate that the JSON is structurally valid by default, unless you're setting the [SkipValidation](https://docs.microsoft.com/dotnet/api/system.text.json.jsonwriteroptions.skipvalidation?view=net-6.0#system-text-json-jsonwriteroptions-skipvalidation) option.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 18]
```diff
@@ -140,5 +141,52 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+            ReadOnlySpan<byte> inputValue = value;
+
+            // We will ignore the leading and trailing occurrences of the white space in the json check.
+            int start = 0;
+            while (start < inputValue.Length && char.IsWhiteSpace(Convert.ToChar(inputValue[start])))
+            {
+                start++;
+            }
+            int end = inputValue.Length - 1;
+            while (end >= start && char.IsWhiteSpace(Convert.ToChar(inputValue[end])))
```
Comment (jsquire) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:156]: Why do we care to trim whitespace?  `Utf8JsonWriter` is going to collapse all whitespace and indentation by default unless we're explicitly setting the [Indented](https://docs.microsoft.com/dotnet/api/system.text.json.jsonwriteroptions.indented?view=net-6.0#system-text-json-jsonwriteroptions-indented) option.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 16 above)
Comment (christothes) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:164]: Taking a step back, is it even valid to store use `BinaryData` to store values that cannot be serialized to json?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 18 above)
Comment (jsquire) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:156]: `Convert.ToChar` just does a straight `(char)` cast; I'd recommend skipping the method call and just doing the cast.

_([src](https://source.dot.net/#System.Private.CoreLib/Convert.cs,05d59fb918111652))_

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs [hunk 19]
```diff
@@ -140,5 +141,52 @@ public static void WriteObjectValue(this Utf8JsonWriter writer, object value)
...
+            int len = end - start + 1;
+
+            // Check if it is possible to serialize the value as json value.
+            // Possible match kinds: Json Object, Json Array, Json string, Json value true, Json value false, Json number.
+            // Our logic here is that, for example, we check if the first and the last bytes are '{' and '}', respectively.
+            if (len > 0 && (
+                inputValue[start] == (byte)'{' && inputValue[end] == (byte)'}' ||
+                inputValue[start] == (byte)'[' && inputValue[end] == (byte)']' ||
+                inputValue[start] == (byte)'"' && inputValue[end] == (byte)'"' ||
+                inputValue[start] == (byte)'t' && inputValue[end] == (byte)'e' ||
```
Comment (jsquire) [inline src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:169]: What if my value is:

- telephone
- fortnite
- 9one1
- ["invalid]
- {"no]}

Unless I'm misreading, these would all pass this validation.

**Comment by jsquire** [comment]
It would be quite helpful if this PR had some context in the description for what benefit this change is intending to provide.   The branch is called "fix-binarydata-serialization" ...  can you help us understand in what way it is broken?

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 16 above)
Comment (christothes) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:164]: oh -  I see now that we handle the edge case in the catch block

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 18 above)
Comment (m-nash) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:156]: We are trying to do a "fast" check to see if we should call WriteJson or WriteString.  If we incorrectly call WriteString for json it comes out as `"{"foo": "bar"}"` with extra quotes around it which is not correct.  We could just try to parse with JsonElement and catch always but this is very slow.  We could also use JsonReader which is bit faster.  Instead we believe we can be 90%+ accurate with this method and therefore only incur the slow parsing fail in a small percentage of cases.  Our worst case runtime is no worse than always doing a full parse, but our average case is significantly better.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 17 above)
Comment (m-nash) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:165]: Good call out on `null`.

Json and String are very simple to deal with, the challenge that we are attempting to handle here are the other primitives namely `null`, `number`, `bool`.

If we want to only use BinaryData for `AnyObject` we can safely just serialize as json which is the current behavior.  If we want to expand its use to `Any` as well we must deal with these other cases also.

Deserialization is easy as no interpretation happens we simply store the raw bytes in the binary data and let the customer choose how to interpret them.  Serialization is where we now need to know how to write out what was given to us.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 19 above)
Comment (m-nash) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:169]: The first three bullets are the same comment [here](https://github.com/Azure/autorest.csharp/pull/2544#discussion_r943715228).  The last two would be handled by the catch block and written as `"["invalid]"` and `"{"no]}"` respectively.  Again the pre-parse is not meant to be 100% accurate but its simply meant to lower our average case by eliminating things that we know for sure are NOT valid json.

File: test/AutoRest.Shared.Tests/WriterExtensionTests.cs (hunk 13 above)
Comment (Yao725) [reply test/AutoRest.Shared.Tests/WriterExtensionTests.cs:137]: I have added a test case in MgmtExtensionResourceTests to see if the result is expected when we first serialize the model, which contains a binary data property and then deserialize it.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 16 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:164]: Yes, we think that this check will be right in most of the cases and use catch to handle the worst case, in which the `BinaryData` passed the check but isn't a valid json. We use this fast check to avoid trying to parse the json everytime and then catch the exception.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 17 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs:165]: Added `null` to the check list.

File: src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs (hunk 18 above)
Comment (Yao725) [reply src/assets/Generator.Shared/Utf8JsonWriterExtensions.cs]: changed, thank you!

**Comment by m-nash** [comment]
@Yao725 based on our findings lets pivot this PR to address the following https://github.com/Azure/autorest.csharp/issues/2447#issuecomment-1217234786 and we will look into what else we can do for documentation on BinaryData itself in the future.

File: src/AutoRest.CSharp/Common/Generation/Writers/ModelWriter.cs [hunk 20]
```diff
@@ -288,13 +288,48 @@ private void BuildHierarchy(ObjectTypeProperty property, Stack<ObjectTypePropert
 
         private FormattableString CreatePropertyDescription(ObjectTypeProperty property, string? overrideName = null)
         {
+            FormattableString binaryDataExtraDescription = Configuration.AzureArm &&
```
Comment (m-nash) [inline src/AutoRest.CSharp/Common/Generation/Writers/ModelWriter.cs:291]: remove the if arm check I believe this should apply to all libraries

**Review by m-nash — APPROVED** [review]

File: src/AutoRest.CSharp/Common/Generation/Writers/ModelWriter.cs (hunk 20 above)
Comment (Yao725) [reply src/AutoRest.CSharp/Common/Generation/Writers/ModelWriter.cs]: removed
